### PR TITLE
feat(perf): Add Performance Overhead section

### LINKED ISSUE #7262: Document overhead of Performance monitoring
### Core or SDK?

Platform/SDK

### Which part? Which one?

SDKs

### Description

We are frequently asked about performance overhead, and while it's hard to put a number to it, we should help our users to understand possible performance implications and what we are doing to mitigate them. 

#### Customer Feedback (paraphrased)
“I don’t need benchmarking statistics. I want to know how the sdk works-under-the hood, and then I can analyze my app and decide if it’s something to be worried about, because only I know the intricacies and architecture of my app”

“I’m scared to let the sentry sdk run because I don’t know what it’s doing. For analogy, if your trusted auto technician said your vehicle needs 5 hours of maintenance work and problem fixing, you’d ask for a description of the required work, just so you’re aware. And then you’d agree (or not). I trust you Sentry but I’m still looking for the same awareness regarding your SDK for Performance Monitoring - what is it doing?”

### Suggested Solution

* Focus on talking about *how the sdk is* *designed* and *works* rather than referencing lab-like benchmarking experiments which are expensive to run.
* Something similar to our existing documentation for perf impact of Profiling and Session Replay, see examples below.
* Something that addresses the concern in the above 2 customer feedback quotes.
* Consider including examples of harmful things the sdk is-not-doing, to assure it's okay to use.

Examples:
* https://docs.sentry.io/product/profiling/performance-overhead mentions "1 to 5%" at the bottom but even if you remove this line, this documentation is still helpful.
* https://docs.sentry.io/product/session-replay/performance-overhead makes 0 mention of a quantified performance impact like "3% CPU" and this is still a helpful page of documentation.
* “trace generation overhead” by [Google’s Dapper Tracing](https://static.googleusercontent.com/media/research.google.com/en//archive/papers/dapper-2010-1.pdf) whitepaper sections 4.1, 4.2, 4.3
* New Relic's [browser monitoring and performance impact](https://docs.newrelic.com/docs/browser/new-relic-browser/performance-quality/browser-monitoring-performance-impact/#:~:text=Network%20impact,unload%20%2C%20pageshow%20%2C%20and%20pagehide%20.) documentation. Notice its general speak about web apps + sdk's. You could remove the "overhead of less than 15ms per page." at the bottom and it is still helpful documentation.
* [UX Cam's](https://developer.uxcam.com/docs/implementation#does-uxcam-affect-the-performance-of-the-application) documentation also does not give CPU/benchmarking statistics on this subject, for their Session Replay product.
* [Datadog's](https://docs.datadoghq.com/real_user_monitoring/session_replay/) documentation does not mention CPU/benchmarking statistics for their Session Replay other than a network bandwidth aspect, "Datadog also reduces the load on a browser’s UI thread by delegating most of the CPU-intensive work (such as compression) to a dedicated web worker. "
* [Fullstory's](https://www.fullstory.com/blog/fullstory-and-page-performance/) documentation divides it into 1. script is small 2. script is edge-cached by CDN to reduce time 3. UI thread utilization 4. Upload bandwdith, and do we do the following at sentry, "FullStory uses Event Handlers that simply add all user interaction events to an in-memory queue to be processed later, rather than processing them in real time. " ?

### Not The Solution / What Is Not Being Asked
Benchmarking experiments, scientific analysis, answers like, “3% of CPU”. The Session Replay docs don’t mention this and the Profiling docs do at the bottom but you could remove that line and it’d still be a useful page of docs.

cc @lizokm

### PR DESCRIPTION
**Pull request by sl0thentr0py**
resolves https://github.com/getsentry/sentry-docs/issues/7262

### CODE REVIEW

File: src/docs/product/performance/performance-overhead.mdx [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+- Incoming and outgoing HTTP requests
+- Background worker tasks
+
+To do this instrumentation, Sentry SDKs wrap or monkeypatch specific functions within popular libraries or frameworks and measure the execution time of these operations. These are then packaged and serialized into a payload containing a [Transaction](/product/performance/transaction-summary/#what-is-a-transaction) (containing Spans) that is transmitted to event ingestion servers via the internet.
+
+The SDKs also make sure to propagate some headers on outgoing requests so that [Distributed Tracing](/product/sentry-basics/tracing/) can work effectively.
+
+<Note>
+
+For most applications, the performance overhead of enabling performance in production is imperceptible to end users.
```
Comment (cleptric) [inline src/docs/product/performance/performance-overhead.mdx:19]: Let's make this more vague so we do not over-promise anything.

File: src/docs/product/performance/performance-overhead.mdx [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+</Note>
+
+## Will Performance break my application?
+
+Each SDK has limits on the data structures used to record all the above data. If your application is doing too many things, we will simply hit those limits and thus avoid scenarios where the SDK itself takes up too much CPU time or grows memory usage unboundedly. We will also capture Outcomes (statistics) whenever these scenarios occur to be transparent about what events we are dropping for what reason.
```
Comment (cleptric) [inline src/docs/product/performance/performance-overhead.mdx:25]: I would clarify that only some SDKs capture client reports (I would also name and link to it).

Comment (sl0thentr0py) [reply src/docs/product/performance/performance-overhead.mdx:27]: client reports are not a user facing concept, only on develop.

File: src/docs/product/performance/performance-overhead.mdx (hunk 1 above)
Comment (sl0thentr0py) [reply src/docs/product/performance/performance-overhead.mdx:21]: what is more vague than this?

File: src/docs/product/performance/performance-overhead.mdx [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+Performance works by instrumenting certain important parts of typical applications such as:
+
+- Database calls
+- Incoming and outgoing HTTP requests
+- Background worker tasks
+
+To do this instrumentation, Sentry SDKs wrap or monkeypatch specific functions within popular libraries or frameworks and measure the execution time of these operations. These are then packaged and serialized into a payload containing a [Transaction](/product/performance/transaction-summary/#what-is-a-transaction) (containing Spans) that is transmitted to event ingestion servers via the internet.
+
+The SDKs also make sure to propagate some headers on outgoing requests so that [Distributed Tracing](/product/sentry-basics/tracing/) can work effectively.
```
[the remark this answers is not in the document]
Comment (sl0thentr0py) [reply src/docs/product/performance/performance-overhead.mdx:18]: this is about overhead, not the technical implementation. Details can go on the tracing page.

**Comment by sl0thentr0py** [comment]
question is if we want to have that on this page. I'll leave the decision to the product people @smeubank and @danielkhan.

File: src/docs/product/performance/performance-overhead.mdx [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+<Note>
+
+For most applications, the performance overhead of enabling performance in production is imperceptible to end users.
+
+</Note>
+
+## Will Performance break my application?
+
+Each SDK has limits on the data structures used to record all the above data. If your application is doing too many things, we will simply hit those limits and thus avoid scenarios where the SDK itself takes up too much CPU time or grows memory usage unboundedly. We will also capture Outcomes (statistics) whenever these scenarios occur to be transparent about what events we are dropping for what reason.
```
Comment (vivianyentran) [inline src/docs/product/performance/performance-overhead.mdx:28]:
```suggestion
Each SDK has limits on the data structures used to record all the above data. If your application is doing too many things, we will simply hit those limits and thus avoid scenarios where the SDK itself takes up too much CPU time or grows memory usage unboundedly. We will also capture outcomes (statistics) whenever these scenarios occur to be transparent about what events we are dropping for what reason.
```

I'm not really sure what outcomes is referring to here... If there's an action item for users to view the statistics somewhere, can we clarify how?

**Review by vivianyentran — APPROVED** [review]
Thanks for adding this doc! +1 that it would be helpful to add a link to the blog post and the web-benchmark page somewhere

File: src/docs/product/performance/performance-overhead.mdx (hunk 4 above)
[the remark this answers is not in the document]
Comment (sl0thentr0py) [reply src/docs/product/performance/performance-overhead.mdx:30]: From the linked issue

> Not The Solution / What Is Not Being Asked
 Benchmarking experiments, scientific analysis, answers like, “3% of CPU”. The Session Replay docs don’t mention this and the Profiling docs do at the bottom but you could remove that line and it’d still be a useful page of docs.

Comment (sl0thentr0py) [reply src/docs/product/performance/performance-overhead.mdx:30]: some of them are exposed to the users, some not, the best we can link to is https://docs.sentry.io/product/stats/

**Comment by sl0thentr0py** [comment]
i'm merging this, we can always add more stuff later

**Pull request merged**